### ISSUE TITLE
How can I use dora for model training?

### ISSUE THREAD

**Issue by npuichigo** [dora-rs/dora#393]
Fantastic work first!

But I cannot find any example on how to use dora for general model training, saying I would like to use dora as a data pipeline for training in PyTorch, for example. I think it's good to provide an toy example like mnist

**Comment by haixuanTao** [comment]
Absolutely! Are you more interested in real-time training? Or you prefer do off-line training within you're regular environment? The ladder should be easier to do using dora-record, as the exported format is arrow so that you can read it with pandas and then do everything with pandas, numpy, torch, tensorflow...

**Comment by npuichigo** [comment]
I'm wondering if dora is the same as ray or maybe there's something special?

**Comment by haixuanTao** [comment]
Are you referring to: https://github.com/ray-project/ray 
In which case, I think they focus on giving a full-stack pipeline for AI Model training and serving. We're more focused on producing real-time AI-based application that can typically run on a car like dora-drives or a small robot like dora-robomaster. 

One of the challenge we aim at tackling is things like integrating sensors, integrating controls, all of this in an environment that is easy to dev with, with feature like hot-reloading, opentelemetry tracing, metrics and logging, and much more :) 

We can do training through dora-record to export the data collected in arrow format. And you can then use your most loved software stack to do the training. This can be ray!

**Comment by heyong4725** [comment]
Closing: dora is not intended as a data pipeline for ML model training. The framework targets real-time, on-device dataflow for robotics and agentic AI applications — the workloads where zero-copy SHM, single-threaded `on_input`, and low-latency message routing actually pay off. That design space does not overlap meaningfully with offline / batch training pipelines, which is what tools like Ray, Lightning, or plain PyTorch DataLoader serve well.

There is one adjacent surface that occasionally trips this question: `dora record` writes captured runtime data to `.drec` files in Arrow format, which can be read back from PyTorch via pyarrow if you want to train a model on data that was *recorded* from a live dora dataflow. That works, but it is a one-way export for offline analysis — not dora acting as the training pipeline itself.

If your use case is "feed a running dataflow into model training in real time" (online / RL-style), dora can be the upstream sensor + processing layer, with a node that hands batches to whatever training framework you are using — but the training itself lives outside dora. Open a fresh issue if that pattern needs better documentation; this question is otherwise answered.
